jetty-util/src/test/java/org/eclipse/jetty/util/BlockingCallbackTest.java

```diff
@@ -0,0 +1,161 @@
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util;
+
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.lessThan;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class BlockingCallbackTest
+{
+    final AtomicInteger notComplete = new AtomicInteger();
+
+    @Test
+    public void testDone() throws Exception
+    {
+        long start;
+        try (BlockingCallback blocker = new BlockingCallback())
+        {
+            blocker.succeeded();
+            start = TimeUnit.NANOSECONDS.toMillis(System.nanoTime());
+        }
+        assertThat(TimeUnit.NANOSECONDS.toMillis(System.nanoTime()) - start, lessThan(500L));
+        assertEquals(0, notComplete.get());
+    }
+
+    @Test
+    public void testGetDone() throws Exception
+    {
+        long start;
+        try (BlockingCallback blocker = new BlockingCallback())
+        {
+            final CountDownLatch latch = new CountDownLatch(1);
+
+            new Thread(new Runnable()
+            {
+                @Override
+                public void run()
+                {
+                    latch.countDown();
+                    try
+                    {
+                        TimeUnit.MILLISECONDS.sleep(100);
+                    }
+                    catch (Exception e)
+                    {
+                        e.printStackTrace();
+                    }
+                    blocker.succeeded();
+                }
+            }).start();
+
+            latch.await();
+            start = TimeUnit.NANOSECONDS.toMillis(System.nanoTime());
+        }
+        assertThat(TimeUnit.NANOSECONDS.toMillis(System.nanoTime()) - start, greaterThan(10L));
+        assertThat(TimeUnit.NANOSECONDS.toMillis(System.nanoTime()) - start, lessThan(1000L));
+        assertEquals(0, notComplete.get());
+    }
+
+    @Test
+    public void testFailed() throws Exception
+    {
+        final Exception ex = new Exception("FAILED");
+        long start = Long.MIN_VALUE;
+        try
+        {
+            try (BlockingCallback blocker = new BlockingCallback())
+            {
+                blocker.failed(ex);
+            }
+            fail("Should have thrown IOException");
+        }
+        catch (IOException e)
+        {
+            start = TimeUnit.NANOSECONDS.toMillis(System.nanoTime());
+            assertEquals(ex, e.getCause());
+        }
+        assertThat(TimeUnit.NANOSECONDS.toMillis(System.nanoTime()) - start, lessThan(100L));
```

Comment:
Why the timing assertion?

Reply:
How else do you make a test flakey???    These tests are copied from SharedBlockingCallback, so I didn't look that closely.    I think they are just checking that something blocked or didn't block... ie didn't take too much time in this case.    I don't think I've seen SharedBlockingCallbackTest as a flakey test, so I think they are pretty reliable.

Reply:
You don't want your tests to ever be flakey, do you? In this case, a simple long GC at the wrong time would make this test fail while there's nothing wrong with the code.

It's just me rambling again each time I see a test that relies on time passing, I think I developed an allergy to them over the years. :-)

Reply:
I forgot the the "/s" for sarcasm.   I agree that such tests are not the best, but I'm not in the mood to rewrite at the moment.